classes/local/store/s3/client.php

```diff
@@ -79,44 +79,91 @@ public function __wakeup() {
         // it will be serialised, so re-retrive them now.
         $config = manager::get_objectfs_config();
         $this->set_client($config);
-        $this->client->registerStreamWrapper();
+        if ($this->is_functional()) {
+            $this->client->registerStreamWrapper();
+        }
     }
 
-    public function set_client($config) {
-        $options = array(
-            'region' => $config->s3_region,
-            'version' => AWS_API_VERSION
-        );
+    /**
+     * Check if the client is functional.
+     * @return bool
+     */
+    private function is_functional() {
+        return !is_null($this->client);
+    }
 
-        if (empty($config->s3_usesdkcreds)) {
-            $options['credentials'] = array('key' => $config->s3_key, 'secret' => $config->s3_secret);
+    /**
+     * Check if the client configured properly.
+     *
+     * @param \stdClass $config Client config.
+     * @return bool
+     */
+    private function is_configured($config) {
+        if (empty($config->s3_bucket)) {
+            return false;
         }
 
-        if ($config->useproxy) {
-            $options['http'] = array('proxy' => $this->get_proxy_string());
+        if (empty($config->s3_region)) {
+            return false;
         }
 
-        // Support base_url config for aws api compatible endpoints.
-        if ($config->s3_base_url) {
-            $options['endpoint'] = $config->s3_base_url;
+        if (empty($config->s3_usesdkcreds)) {
+            if (empty($config->s3_key) && empty($config->s3_secret)) {
+                return false;
+            }
         }
 
-        $this->client = \Aws\S3\S3Client::factory($options);
+        return true;
+    }
+
+    /**
+     * Set the client.
+     *
+     * @param \stdClass $config Client config.
+     */
+    public function set_client($config) {
+        if (!$this->is_configured($config)) {
+            $this->client = null;
+        } else {
+            $options = array(
+                'region' => $config->s3_region,
+                'version' => AWS_API_VERSION
+            );
+
+            if (empty($config->s3_usesdkcreds)) {
+                $options['credentials'] = array('key' => $config->s3_key, 'secret' => $config->s3_secret);
```

Comment:
I prefer square brackets for arrays, but this being on multiple lines makes more sense imo. And multi line arrays are better using the square bracket form because of the ability to use a trailing comma without potential compatibility disaster :smile: 
```suggestion
                $options['credentials'] = [
                    'key' => $config->s3_key,  
                    'secret' => $config->s3_secret,
                ];
```

Reply:
100% agree. But, formatting should be fixed in a different PR.